### ISSUE TITLE
Follow token restriction disallows keywords after path fragment specifier (and likely others)

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust#101788]
I just wanted to add match guard support to the [as_variant](https://docs.rs/as_variant/latest/as_variant/) crate I published yesterday, but found that it was impossible without turning the macro into a token muncher (something I'm not familiar enough with and also find rather unappealing).

Basically, I tried this:

```rust
macro_rules! as_variant {
    ( $enum:expr, $( $variants:path )|* $( if $guard:expr )? ) => { /* ... */ };
}
```

and got the error

```
error: `$variants:path` may be followed by `if`, which is not allowed for `path` fragments
   --> src/lib.rs:120:44
    |
120 |     ( $enum:expr, $( $variants:path )|* $( if $guard:expr )? ) => {
    |                                            ^^ not allowed after `path` fragments
    |
    = note: allowed there are: `{`, `[`, `=>`, `,`, `>`, `=`, `:`, `;`, `|`, `as` or `where`
```

which seems wrong.

Actually after writing most of this report and seeing `as` and `where` in the list I found a workaround that's not token munching, I can use `pat_param` instead of `path`, but I feel like what I wrote originally should still be allowed.

**Comment by jplatte** [comment]
Ah, that's unfortunate but understandable. Should I close the issue then?

**Comment by jplatte** [comment]
Right, basically I meant this issue as a feature request to add `if` and possibly other keywords to the follow-set of `path` (and maybe `ty`, apparently those use the same set).
